### PR TITLE
8301116: Parallelize TLAB resizing in G1

### PR DESCRIPTION
**Pull request by tschatzl**
Hi all,

  can I get reviews for this change that moves TLAB resizing into the second parallel post evacuation phase?

The change is fairly straightforward except maybe for the following:
* there is a new claimer for `JavaThreads` because the `Threads::possibly_parallel_threads_do` method to parallelize does not work well with very little per-thread work. Actually with a moderate amount of threads (~20) performance would already be many times slower than doing things serially.
* moving the TLAB resizing out of `G1CollectedHeap::prepare_tlabs_for_mutator` made that method and the enclosing timing a bit obsolete, so I repurposed it as a "do preparatory work for the mutator during young gc" phase. That resulted in minor cleanup and regularization of what is done during that phase (wrt to what the corresponding method for full gc does).

Testing: gha, local perf testing, tier1

Thanks,
  Thomas

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8301116](https://bugs.openjdk.org/browse/JDK-8301116): Parallelize TLAB resizing in G1

### Reviewers
 * [Albert Mingkun Yang](https://openjdk.org/census#ayang) (@albertnetymk - **Reviewer**) ⚠️ Review applies to [6bff3cc1](https://git.openjdk.org/jdk/pull/12360/files/6bff3cc17a17e0d9ec282c50872b4a56e297b68a)
 * [Ivan Walulya](https://openjdk.org/census#iwalulya) (@walulyai - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12360/head:pull/12360` \
`$ git checkout pull/12360`

Update a local copy of the PR: \
`$ git checkout pull/12360` \
`$ git pull https://git.openjdk.org/jdk pull/12360/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12360`

View PR using the GUI difftool: \
`$ git pr show -t 12360`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12360.diff">https://git.openjdk.org/jdk/pull/12360.diff</a>

</details>

### CODE REVIEW

**Comment by tschatzl** [comment]
The reason for this change (and the follow-up(s)) is that in applications with a large number of Java threads (20k+), TLAB retirement, log buffer flushing and TLAB resizing takes 3-7% of total pause time. 
Parallelization can reduce this time significantly.

File: src/hotspot/share/runtime/threadSMR.hpp [hunk 1]
```diff
@@ -201,7 +201,7 @@ class ThreadsList : public CHeapObj<mtThread> {
 
   JavaThread *const thread_at(uint i) const { return _threads[i]; }
 
-  JavaThread *const *threads() const        { return _threads; }
```
Comment (dholmes-ora) [inline src/hotspot/share/runtime/threadSMR.hpp:204]: Seems an unrelated change. ??

Comment (tschatzl) [reply src/hotspot/share/runtime/threadSMR.hpp:204]: No, the operations on the `JavaThread` during these phases change the `JavaThread` state (actually, the referenced TLAB state).

E.g. `g1YoungGCPostEvacuateTasks.cpp:720`:
```
void do_work(uint worker_id) override {
    JavaThread* const* list;
    uint count;
    while ((list = _claimer.claim(count)) != nullptr) {
      for (uint i = 0; i < count; i++) {
        list[i]->tlab().resize();  <--- here, the tlab needs to be modified (but not the JavaThread)
      }
    }
  }
```

I could `const_cast` that one away, but that seemed really ugly to me.

In a follow-up change there is a similar access to the threads' TLAB. Maybe there is a better way to do that (apart from `const_cast`ing).

Comment (tschatzl) [reply src/hotspot/share/runtime/threadSMR.hpp:204]: To be more exact, the removal of the const happens in `g1CollectedHeap.inline.hpp:124`:
```
inline JavaThread* const* G1JavaThreadsListClaimer::claim(uint& count) {
  count = 0;
  if (Atomic::load(&_cur_claim) >= _list.length()) {
    return nullptr;
  }
  uint claim = Atomic::fetch_and_add(&_cur_claim, _claim_step);
  if (claim >= _list.length()) {
    return nullptr;
  }
  count = MIN2(_list.length() - claim, _claim_step);
  return _list.list()->threads() + claim;                         <--- here
}
```
because of the mentioned access in `g1YoungGCPostEvacuateTasks.cpp:720`.

Comment (tschatzl) [reply src/hotspot/share/runtime/threadSMR.hpp:204]: Using `Threads::possibly_parallel_oops_do`, thread iteration/claiming seems to be the bottleneck, i.e. claiming the token.

One issue is that all threads need to traverse the array from the beginning to get to the current claim position - so the minimum processing time for a single thread is iterating the complete thread array and checking all tokens for it.  That does not matter so much if the work per thread is big (like when walking the stacks for oops - but even then I think it would be noticable, need to check).

The other is that with little work per thread threads like in this situation they seem to be contending heavily on the JavaThread claim tokens, so this "parallelization" is quite a pessimization - I've measured ~4x slower with 18 threads than using the single-threaded version.

Comment (tschatzl) [reply src/hotspot/share/runtime/threadSMR.hpp:204]: It looks like this has been a leftover of some older version (the `G1JavaThreadClaimer` is fairly "new") - I will remove this change.
Thanks for making me look at this again in detail!

Comment (kimbarrett) [reply src/hotspot/share/runtime/threadSMR.hpp]: > To be more exact, the removal of the const happens in `g1CollectedHeap.inline.hpp:124`:
> 
> ```
> inline JavaThread* const* G1JavaThreadsListClaimer::claim(uint& count) {
>   count = 0;
>   if (Atomic::load(&_cur_claim) >= _list.length()) {
>     return nullptr;
>   }
>   uint claim = Atomic::fetch_and_add(&_cur_claim, _claim_step);
>   if (claim >= _list.length()) {
>     return nullptr;
>   }
>   count = MIN2(_list.length() - claim, _claim_step);
>   return _list.list()->threads() + claim;                         <--- here
> }
> ```
> 
> because of the mentioned access in `g1YoungGCPostEvacuateTasks.cpp:720`.

I think a better solution might be to declare the `Thread::_tlab` member `mutable` and `Thread::tlab() const`.

Comment (kimbarrett) [reply src/hotspot/share/runtime/threadSMR.hpp]: > Using `Threads::possibly_parallel_oops_do`, thread iteration/claiming seems to be the bottleneck, i.e. claiming the token.
> 
> One issue is that all threads need to traverse the array from the beginning to get to the current claim position - so the minimum processing time for a single thread is iterating the complete thread array and checking all tokens for it. That does not matter so much if the work per thread is big (like when walking the stacks for oops - but even then I think it would be noticable, need to check).
> 
> The other is that with little work per thread threads like in this situation they seem to be contending heavily on the JavaThread claim tokens, so this "parallelization" is quite a pessimization - I've measured ~4x slower with 18 threads than using the single-threaded version (on ~21k JavaThreads)

Looking at `Threads::possibly_parallel_threads_do`, I'm thinking it could use a redesign to use the technique
being used in `G1JavaThreadsListClaimer`.  The existing design and implementation dates from when the
threads-list was an actual linked list, rather than the newer ThreadsList.  But maybe that's future work.

Comment (tschatzl) [reply src/hotspot/share/runtime/threadSMR.hpp]: > > Using `Threads::possibly_parallel_oops_do`, thread iteration/claiming seems to be the bottleneck, i.e. claiming the token.
> > One issue is that all threads need to traverse the array from the beginning to get to the current claim position [...]
> > The other is that with little work per thread threads like in this situation they seem to be contending heavily on the JavaThread claim tokens [...]
> 
> Looking at `Threads::possibly_parallel_threads_do`, I'm thinking it could use a redesign to use the technique being used in `G1JavaThreadsListClaimer`. The existing design and implementation dates from when the threads-list was an actual linked list, rather than the newer ThreadsList. But maybe that's future work.

I considered that, but the non-java threads still use the linked list design, so I moved that to future work :) I can investigate that a bit more though, but would like to keep it as is for this change.

**Review by albertnetymk — APPROVED** [review]
Two minor suggestions:

1. I think "inlining" the logic of `G1JavaThreadsListClaimer` makes the flow in the caller more transparent.

2. `ResizeThreadLABs` -> `ResizeTLABs`, just to be consistent with other *TLAB occurrences.

File: src/hotspot/share/runtime/threadSMR.hpp (hunk 1 above)
Comment (tschatzl) [reply src/hotspot/share/runtime/threadSMR.hpp]: > I think a better solution might be to declare the `Thread::_tlab` member `mutable` and `Thread::tlab() const`.

The code works as is without the removed `const` after all, so I'm not sure I should do this suggested change.

**Review by walulyai — APPROVED** [review]
Lgtm!

**Comment by tschatzl** [comment]
I'm pushing this because I consider @dholmes-ora and @kimbarrett 's question resolved - the code in question has been reverted after all.

Thanks everyone for your reviews. 

/integrate
